packages/bp/src/core/dialog/dialog-engine.ts

```diff
@@ -219,79 +268,88 @@ export class DialogEngine {
     if (!timeoutNode) {
-      timeoutNode = findNodeWithoutError(currentFlow, 'timeout')
+      timeoutNode = this.findNodeWithoutError(botId, currentFlow, 'timeout')
     }
 
     // Check for a timeout property in the current flow
     if (!timeoutNode) {
       const timeoutNodeName = _.get(timeoutFlow, 'timeoutNode')
       if (timeoutNodeName) {
-        timeoutNode = findNodeWithoutError(timeoutFlow, timeoutNodeName)
+        timeoutNode = this.findNodeWithoutError(botId, timeoutFlow, timeoutNodeName)
       }
     }
 
     // Check for a timeout.flow.json and get the start node
     if (!timeoutNode) {
-      timeoutFlow = findFlowWithoutError('timeout.flow.json')
+      timeoutFlow = this.findFlowWithoutError(botId, 'timeout.flow.json')
       if (timeoutFlow) {
         const startNodeName = timeoutFlow.startNode
-        timeoutNode = findNodeWithoutError(timeoutFlow, startNodeName)
+        timeoutNode = this.findNodeWithoutError(botId, timeoutFlow, startNodeName)
       }
     }
 
     if (!timeoutNode || !timeoutFlow) {
       throw new TimeoutNodeNotFound(`Could not find any timeout node or flow for session "${sessionId}"`)
     }
 
-    if (event.state.context) {
-      event.state.context.currentNode = timeoutNode.name
-      event.state.context.currentFlow = timeoutFlow.name
-      event.state.context.queue = undefined
-      event.state.context.previousFlow = currentFlow.name
-      event.state.context.previousNode = currentNode?.name as string
-      event.state.context.jumpPoints = [
-        ...(event.state.context?.jumpPoints || []),
-        {
-          flow: currentFlow.name,
-          node: currentNode?.name as string
-        }
-      ]
-      event.state.context.hasJumped = true
-    }
+    this.fillContextForTransition(event, { currentFlow, currentNode, nextFlow: timeoutFlow, nextNode: timeoutNode })
 
     return this.processEvent(sessionId, event)
   }
 
-  private _endFlow(event: IO.IncomingEvent) {
-    event.state.context = {}
-    event.state.temp = {}
+  public async processConversationEnd(botId: string, sessionId: string, event: IO.IncomingEvent) {
+    this._debug(event.botId, event.target, 'processing conversation end')
+
+    await this._loadFlows(botId)
+
+    const currentFlow = this._findFlow(botId, event.state.context?.currentFlow!)
+    const currentNode = this.findNodeWithoutError(botId, currentFlow, event.state.context?.currentNode)
+
+    // Check for a conversation_end.flow.json
+    const conversationEndFlow = this.findFlowWithoutError(botId, 'conversation_end.flow.json')
+
+    // Don`t trigger on timeout or if it is already at the conversation end flow
+    if (
+      !conversationEndFlow ||
+      currentFlow.name === 'conversation_end.flow.json' ||
+      currentFlow.name === 'timeout.flow.json'
+    ) {
+      return true
+    }
+
+    const api = await createForGlobalHooks()
+    await this.hookService.executeHook(new Hooks.BeforeConversationEnd(api, event))
```

Comment:
Is there a case where a user would want a hook without having a "conversation_end" flow ? Maybe a user could jump the user to a specific flow (using jumpTo) ? Or is the hook specifically for that conversation_end flow ?

Reply:
Changed to support the following scenarios:
No Flow / Hook Active
Flow Active/ No Hook
Flow Active / Hook Active
Flow Active / Hook Active With transition (possible loop) (will skip the flow for the previous conversation end event) 
Flow Active With transition (possible loop)/ Hook Active